### PR TITLE
DO NOT MERGE: Proposal for IS-07 inclusion in IS-04

### PR DESCRIPTION
**Pull request by andrewbonney**
Another revision of IS-07 proposals for inclusion in IS-04. This one is written with a few things in mind:

* In the future, additional attributes should be defined in the parameter registers to avoid version bumps. All of the changes are additive as a result and wouldn't require explicit schema changes to accommodate (due to additionalProperties: true). In this case the schema changes are included to be explicit.
* Ensures all IS-07 data can be observed at the Flow level to avoid confusion between IS-07 and ST.291 at the Source level. The Source level is instead reserved for additional signalling of commonality between the Flows which sit under it.
* Indicates how an IS-07 Source may have multiple Flows in line with the ID/Timing model in the case of a temperature measurement, where the same measurement but with different units (C/F) can be viewed as equivalent.

Note that as there may be future application/json Flows of other types, the Flow is only an IS-07 one if it includes the 'event_type' attribute.

### CODE REVIEW

File: examples/nodeapi-receivers-get-200.json [hunk 1]
```diff
@@ -21,5 +21,31 @@
...
+    {
+        "description": "",
+        "format": "urn:x-nmos:format:data",
+        "tags": {},
+        "caps": {
+            "media_types": [
+                "application/json"
+            ],
+            "event_types": [
+                "number/temperature",
```
Comment (garethsb) [inline examples/nodeapi-receivers-get-200.json]: Should be "number/temperature/*"?

**Comment by garethsb** [comment]
Are two new schemas missing? "source_data.json" and "flow_json_data.json"?

**Comment by andrewbonney** [comment]
Yes... Fixed

File: examples/nodeapi-receivers-get-200.json (hunk 1 above)
Comment (andrewbonney) [reply examples/nodeapi-receivers-get-200.json]: Yes, quite possibly given the existing definition within IS-07. I'll make that change. The same probably goes for what I've done on the Source side.

**Comment by garethsb** [comment]
Does it bother you that something as generic sounding as an "application/json" data Source or Flow should have something as specific as an (IS-07) "event_type" property?

I was wondering if we could use something more generic, similar to IANA media type sub-parameters (like "charset", or the "sampling" and "colorimetry" ones of "video/raw"), of which the proposed/offered "event_type" were one example? Perhaps I'm grasping for something that's more of a direct counterpart to "cap(abilitie)s" on Receivers?

**Comment by andrewbonney** [comment]
> Does it bother you that something as generic sounding as an "application/json" data Source or Flow should have something as specific as an (IS-07) "event_type" property?

The schemas intentionally don't stop you having 'application/json' without an 'event_type', removing the link to IS-07. It bothered me more when the 'event_type' was only in the Source, as comparing an IS-07 Flow to another JSON based Flow would have been more involved.

One previous PR considered introducing a 'data_type' or similar which could explicitly indicate IS-07 vs. another Source/Flow construction, but I was hoping we could avoid another formal register of types.

Very happy to consider further revisions on the theme. It's been a tricky one to come up with a clear and extensible approach for.

**Comment by garethsb** [comment]
> The schemas intentionally don't stop you having 'application/json' without an 'event_type', removing the link to IS-07.

Yeah, I got that. It's just a spec dependency that seems bothersome.

And invites the usual forward-looking questions like _"What will we do when other specs add other 'application/json' Flows and want to indicate their "fizzbuzz_type"?"_

> One previous PR considered introducing a 'data_type' or similar which could explicitly indicate IS-07 vs. another Source/Flow construction, but I was hoping we could avoid another formal register of types.

Yes, PR #1, I remember...

I guess I was wondering whether IS-07 should define the media type "application/nmos-event-tally+json", with a (IANA-registered?) sub-parameter, "event_type".

Parameters could be embedded in the Flow "media_type" as e.g. "application/nmos-event-tally+json;event_type=number/temperature/C", though that raises the usual naïve string-matching concerns, or separated out like Receiver `"caps"`, something like `"params": { "event_type": "number/temperature/C" }`.

**Comment by andrewbonney** [comment]
> Yeah, I got that. It's just a spec dependency that seems bothersome.
> 
> And invites the usual forward-looking questions like _"What will we do when other specs add other 'application/json' Flows and want to indicate their "fizzbuzz_type"?"_

Including the schemas for this in IS-04 isn't technically required as long as we can agree on the approach. They could instead be stored in the parameter registers, with the intent being that any future Source/Flow type could be added in the same way if it doesn't conflict with something already in existence. Continuing to produce new versions of IS-04 indefinitely to support new use cases like this seems unnecessary.

> I guess I was wondering whether IS-07 should define the media type "application/nmos-event-tally+json", with a (IANA-registered?) sub-parameter, "event_type".

That certainly seems like it could work in principal. As you say the usual string matching concerns come up, particularly if there are multiple sub-parameters in any order. There could also be cases where implementers of existing media_types include sub-parameters as a result, which could again confuse some clients. 'params' does seem cleaner, but somewhat raises the question of why it needs to be in its own object when existing Flow types like video have parameters at the top level in the Flow object. The reason may of course simply be that we'd do it that way for everything (other than label, description etc) if we were to do things from scratch.

**Comment by garethsb** [comment]
> somewhat raises the question of why it needs to be in its own object when existing Flow types like video have parameters at the top level in the Flow object. 

That is a good point...

I guess it made sense to me that there was a specific schema with specific video-focused parameters when the media_type was "video/xxx", and likewise for audio. I felt that "application/json" and "event_type" had a more tenuous connection, but it's only a matter of degree.

**Comment by andrewbonney** [comment]
Thinking on the use of media_type sub-parameters a little more, this may be a little tricky to fit into the model at the Source level (dependent on how you view the parameters' scope). The Source is intended to be agnostic to the media_type, but it's useful to be able to signal the commonality between Flows of a Source at the Source level. This may include a shared 'event_type' between Flows.

**Comment by garethsb** [comment]
Yes, that makes sense. So not media type sub-parameters then.

Which clarifies that "event_type" is basically associated with the format "urn:x-nmos:format:data" rather than a particular media type like "application/json", that just happens to be the only media type we know of that carries events, if that makes sense?

So I now agree that "event_type" is enough like audio "channels" to appear at the top level of Sources, and "event_type" like "sample_rate" and "colorspace" at the top level of Flows. Thanks for walking me through it!

Though it's potentially going to be the first specific capability (after the generically useful "media_types"), adding "event_types" to Receiver "caps" also makes sense in the light of PR #6. I would like us to think about how programmatic matching between Receiver "caps" and Source/Flow properties could be achieved, but that's more in the scope of #6.

**Comment by andrewbonney** [comment]
Yeah, agreed. I'd like to re-visit the 'caps' structure in relation to IS-05 constraints style formatting, but as you say that relates to #6

**Comment by garethsb** [comment]
Ah, I realise one thing that would help with the logic of where "event_type" exists in the Flow schemas, and that would simply be renaming flow_json_data.json to flow_events_data.json (and changing the title/description from "JSON based Flow resource" to "An Events data Flow"). This would indicate that, like flow_sdianc_data.json which introduces "DID_SDID", this is not entirely about the media_type (SDI ANC data might be carried by something other than video/smpte291, like events might be carried by something other than application/json), but rather about something more broad.

**Comment by andrewbonney** [comment]
Just updating here as we discussed elsewhere. This change would require an additional schema for 'flow_json_data' in addition to 'flow_events_data' to account for other uses of 'application/json' (or another similar re-jigging of the schemas).

**Pull request closed without merging**